qiskit/utils/units.py

```diff
@@ -25,16 +32,120 @@ def apply_prefix(value: float, unit: str) -> float:
     Returns:
         Converted value.
 
+    Note:
+        This may induce tiny value error due to internal representation of float object.
+        See https://docs.python.org/3/tutorial/floatingpoint.html for details.
+
     Raises:
-        Exception: If the units aren't recognized.
+        Exception: If the ``units`` aren't recognized.
+        TypeError: If the ``value`` is not a number.
     """
-    downfactors = {"p": 1e12, "n": 1e9, "u": 1e6, "µ": 1e6, "m": 1e3}
-    upfactors = {"k": 1e3, "M": 1e6, "G": 1e9}
-    if not unit:
+    prefactors = {
+        "f": -15,
+        "p": -12,
+        "n": -9,
+        "u": -6,
+        "µ": -6,
+        "m": -3,
+        "k": 3,
+        "M": 6,
+        "G": 9,
+        "T": 12,
+        "P": 15,
+    }
+
+    if not unit or len(unit) == 1:
+        # for example, "m" can represent meter
         return value
-    if unit[0] in downfactors:
-        return value / downfactors[unit[0]]
-    elif unit[0] in upfactors:
-        return value * upfactors[unit[0]]
+
+    if unit[0] not in prefactors:
+        raise Exception(f"Could not understand unit: {unit}")
+
+    try:
+        value = float(value)
+    except TypeError as ex:
+        if isinstance(value, ParameterExpression) and value.parameters:
+            return pow(10, prefactors[unit[0]]) * value
+        raise TypeError(f"Input value {value} is not a number.") from ex
+
+    # to avoid round-off error of prefactor
+    _value = Decimal.from_float(value).scaleb(prefactors[unit[0]])
```

Comment:
The decimal form doesn't really avoid any more round-off errors than just doing it in floating point, if we take care to only work with positive powers of 10.  All positive powers of 10 up to 10^22 have exact representations (`float(1e23)` differs from the exact value by 2^23), and since floating-point operations guarantee that the result will be the closest float to the infinite-precision answer, there's the same amount of rounding here (conversion to float at the end) as there would be if we did the previous multiply-or-divide trick to keep exact representations.

That said, I don't really mind one way or the other that much.  I doubt (hope, even!) that this isn't performance-critical code (where avoiding `Decimal` would be better), and as you say, we'll never really be able to entirely sensibly handle cases such as `apply_prefix(1e-7, 'µs')`, since `float('1e-7') < Decimal('1e-7')`, but `float('1e-1') > Decimal('1e-1')`.

Reply:
Umm... this is good point. If we avoid `Decimal` we don't need to handle exception for `ParameterExpression`. I'll try multiply-or-divide trick too.

Reply:
That worked. https://github.com/Qiskit/qiskit-terra/pull/6885/commits/1ba3e95770076f2afc7c3c88a4cf43cf8c740e16